scripts/merge_profiles.R

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env Rscript
+
+library(optparse)
+library(data.table)
+library(this.path)
+library(parallel)
+
+# Include common utility functions
+source(this.path::here('include', 'utils.R'))
+
+parser = OptionParser()
+parser = add_option(parser, c("-t", "--threads"), type="integer", default=4, help="number of threads [default: %default]")
+parser = add_option(parser, c("-m", "--memory"), type="integer", default=10, help="maximum memory to be used")
+parser = add_option(parser, c("-o", "--out"), type="character", default=NULL, help="output file")
+parser = add_option(parser, c("-i", "--input"), type="character", default=NULL, help="files to be combined")
+parser = add_option(parser, c("-k", "--keys"), type="character", default=NULL, help="columns shared across files and must be used for join")
+parser = add_option(parser, c("-z", "--zeroes"), action="store_true", default=FALSE, help="flag whether there are zeroes in the profile output that should be removed")
+
+opt = parse_args(parser)
+
+threads_n = opt$threads
+memory_lim = opt$memory
+out_file = opt$out
+in_files = as.list(strsplit(opt$input, ",", fixed=TRUE))[[1]]
+keys = as.vector(strsplit(opt$keys, ",", fixed=TRUE))[[1]]
+zeroes = opt$zeroes
+
+setDTthreads(threads = threads_n)
+
+get_dt_from_file <- function(filename, index_keys=c('ID'), remove_zeroes) {
+    # Load libraries, for parallel
+    library(data.table)
+
+    # Read file
+    logmsg("FILE: ", filename)
+    logmsg("  Reading")
+    dt = fread(filename, header=T, data.table=TRUE, sep='\t')
+
+    if (remove_zeroes == TRUE) {
+        # Index by sample data
+        logmsg("  Indexing table by data")
+        sample_col = setdiff(colnames(dt), index_keys)
+        setkeyv(dt, sample_col)
+
+        # Remove 0-values
+        logmsg("  Removing 0's ")
+        logmsg("    before = ", nrow(dt))
+        dt = dt[!J(0),]
+        logmsg("     after = ", nrow(dt))
+    }
+
+    # Index by index_keys
+    logmsg("  Indexing table by ", index_keys)
+    setkeyv(dt, index_keys)
+
+    return(dt)
+}
+
+# ID gets first key status.
+# Others are sorted and sent as secondary keys.
+
+secondary_keys = sort(setdiff(keys, c('ID')))
+merged_dt = NULL
+
+# Get list of DT from list of filenames
+logmsg("Reading ", length(in_files), " files using ", threads_n, " threads")
+cluster <- makeCluster(threads_n)
+clusterExport(cluster, varlist=c("logmsg"))
+dt_list = parLapply(cluster, in_files, get_dt_from_file, index_keys=c('ID', secondary_keys), remove_zeroes=zeroes)
```

Comment:
Would it not be slightly faster to only remove the full-zero rows from the merged table, instead of removing rows from the indiv. inputs and then re-adding them on merge?
I don't see where memory_lim variable is used in the script.

Reply:
You might be right. Let me try to benchmark the `--zeroes` mode.

Reply:
Tested merging 20 `bed.gz` files with `12,696,451` lines each.
### Current implementation:
```
# 2024-06-06 18:54:46 - Reading 20 files using 10 threads
# 2024-06-06 18:56:05 -   done
# 2024-06-06 18:56:05 - Merging
# 2024-06-06 19:00:21 -   done
# 2024-06-06 19:00:21 - Writing merged output
# 2024-06-06 19:00:23 -   done
real    5:39.00
user    801.23
system  24.18
avgmem  0KB
maxmem  7163812KB
cpu     243%
```
### New implementation removing zeroes at the end (identical output):
```
# 2024-06-07 17:19:24 - Reading 20 files using 10 threads
# 2024-06-07 17:24:48 -   done
# 2024-06-07 17:24:48 - Merging
# 2024-06-07 17:26:23 -   done
# 2024-06-07 17:26:23 - Removing rows with all 0's
# 2024-06-07 17:26:23 -     before = 12696451
# 2024-06-07 17:26:27 -      after = 10864149
# 2024-06-07 17:26:51 - Writing merged output
# 2024-06-07 17:26:53 -   done
real    7:34.69
user    304.40
system  45.30
avgmem  0KB
maxmem  11658456KB
cpu     76%
```

It took significantly longer to load the files into list of `data.table` but merging was much faster. Overall it was slower. Perhaps I should try with more files - new implementation could get an edge there. Will try that.

Reply:
For 8 files, it is `2:00min` for old implementation and `3:25min` for new implementation.